substrate/frame/contracts/src/exec.rs

```diff
@@ -1046,6 +1080,12 @@ where
 			Err(error) => (false, Err(error.into())),
 		};
 
+		if success {
+			self.transient_storage.commit_transaction();
+		} else {
+			self.transient_storage.rollback_transaction();
+		}
+
```

Comment:
can we adopt an API similar to the transaction API here, wrapping the code in a closure, that handle start_transaction & comit or rollback, this should make it easier not to introduce bug when we refactor this code.

Reply:
@pgherveou do you mean something like this:
```
		self.transient_storage.with_transaction(|| ->TransactionOutcome<Result<_, DispatchError>> {
			storage::with_transaction(|| -> TransactionOutcome<Result<_, DispatchError>> {
				let output = do_transaction();
```
With this, I encountered issues with the borrow checker. The issue is with `do_transaction()`closure which borrows `self` .